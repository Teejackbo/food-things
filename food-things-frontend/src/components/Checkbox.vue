<template>
  <div
    :class="{ checked }"
    class="checkbox"
    tabindex="0"
    @click="toggle"
    @keyup.space="toggle"
  >
    {{ checked ? 'âœ“' : '' }}
  </div>
</template>

<script>
export default {
  name: 'Checkbox',

  data() {
    return {
      checked: this.value || false,
    }
  },

  props: {
    value: {
      type: Boolean,
      required: false,
      default: false,
    },
  },

  methods: {
    toggle() {
      this.checked = !this.checked
      this.$emit('input', this.checked)
    },
  },
}
</script>

<style lang="scss">
@import '../style/variables';

.checkbox {
  align-items: center;
  background: $grey;
  border-radius: 5px;
  box-shadow: 0px 2px 2px $shadow;
  color: $grey-light;
  display: flex;
  font-size: 21px;
  height: 28px;
  justify-content: center;
  margin: 0.6em 0;
  padding: 0.05em;
  transition: background 0.15s;
  user-select: none;
  width: 28px;

  &.checked {
    background: $primary;
  }
}
</style>
