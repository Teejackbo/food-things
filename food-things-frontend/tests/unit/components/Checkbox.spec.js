import { shallowMount } from '@vue/test-utils'
import Checkbox from '@/components/Checkbox.vue'

describe('<Checkbox />', () => {
  it('Should match snapshot.', () => {
    const wrapper = shallowMount(Checkbox)

    expect(wrapper.element).toMatchSnapshot()
  })

  it('Should match snapshot when value is true.', () => {
    const wrapper = shallowMount(Checkbox, {
      propsData: {
        value: true,
      },
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('Should render a div with the class of checkbox.', () => {
    const wrapper = shallowMount(Checkbox)

    expect(wrapper.contains('div.checkbox')).toBe(true)
  })

  it('Should add the class checked if a true value prop is passed.', () => {
    const wrapper = shallowMount(Checkbox, {
      propsData: {
        value: true,
      },
    })

    expect(wrapper.contains('div.checkbox.checked')).toBe(true)
  })

  it('Should not have the class checked if a false value prop is passed.', () => {
    const wrapper = shallowMount(Checkbox, {
      propsData: {
        value: false,
      },
    })

    expect(wrapper.contains('div.checkbox.checked')).toBe(false)
  })

  it('Should render a tick if it is checked, or nothing if not.', () => {
    const wrapper = shallowMount(Checkbox, {
      propsData: {
        value: true,
      },
    })

    expect(wrapper.text()).toBe('âœ“')
    wrapper.setData({ checked: false })
    expect(wrapper.text()).toBe('')
  })

  it('Should emit an event if it is clicked or space is used to toggle it.', () => {
    const wrapper = shallowMount(Checkbox)

    wrapper.trigger('click')
    expect(wrapper.emitted()).toEqual({ input: [[true]] })

    wrapper.trigger('keyup.space')
    expect(wrapper.emitted()).toEqual({ input: [[true], [false]] })
  })

  it('Clicking or pressing space should toggle the checked class.', () => {
    const wrapper = shallowMount(Checkbox)

    wrapper.trigger('click')
    expect(wrapper.contains('div.checkbox.checked')).toBe(true)

    wrapper.trigger('keyup.space')
    expect(wrapper.contains('div.checkbox.checked')).toBe(false)
  })
})
