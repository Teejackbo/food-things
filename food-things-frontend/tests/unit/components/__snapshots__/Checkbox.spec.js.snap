// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> Should match snapshot when value is false. 1`] = `
<div
  class="checkbox"
  tabindex="0"
>
  
  

</div>
`;

exports[`<Checkbox /> Should match snapshot when value is true. 1`] = `
<div
  class="checkbox checked"
  tabindex="0"
>
  
  âœ“

</div>
`;
